### PR TITLE
topic7

### PR DESCRIPTION
**Pull request by matihaure**

### CODE REVIEW

File: topic7/pom.xml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project
+  xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>netgloo</groupId>
+  <artifactId>spring-boot-mysql-springdatajpa-hibernate</artifactId>
```
Comment (amilcar-infante) [inline topic7/pom.xml:9]: This POM was generated by your environment or did you use some POM from a sample?

Comment (matihaure) [inline topic7/pom.xml:9]: POM from sample

2015-03-05 18:41 GMT-03:00 amilcar-infante notifications@github.com:

File: topic7/src/main/java/controllers/CourseController.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+@Controller
+public class CourseController {
+
+	@Autowired
+	private CourseDao _CourseDao;
+
+	@RequestMapping("/create")
+	@ResponseBody
+	public String create(String name) {
+		try {
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:20]: It's a good practice use try-catch to avoid to un-catch possible errors in your code, but keep in mind that if you are implementing a web service it always need to return a message to FE indicating the result of the operation, in that case an issue (it is possible using an ExceptionHandler)

File: topic7/src/main/java/controllers/CourseController.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+			Course Course = new Course(name);
+			_CourseDao.save(Course);
+		} catch (Exception ex) {
+			return "Error creating the user: " + ex.toString();
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:24]: I see, you are returning a single string :)

File: topic7/src/main/java/controllers/CourseController.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+			return "Error deleting the Course:" + ex.toString();
+		}
+		return "Course succesfully deleted!";
+	}
+
+	@RequestMapping("/findAll")
+	@ResponseBody
+	public String findAll() {
+		try {
+			_CourseDao.findAll();
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:45]: In this method the request should be returning all the courses

File: topic7/src/main/java/controllers/CourseController.java (hunk 3 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/CourseController.java:24]: lol haha

2015-03-05 18:46 GMT-03:00 amilcar-infante notifications@github.com:

File: topic7/src/main/java/controllers/CourseController.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	@RequestMapping("/create")
+	@ResponseBody
+	public String create(String name) {
+		try {
+			Course Course = new Course(name);
+			_CourseDao.save(Course);
+		} catch (Exception ex) {
+			return "Error creating the user: " + ex.toString();
+		}
+		return "Course succesfully created!";
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:26]: A good option in a create method is to return the object just created or a boolean indicating the result

File: topic7/.springBeans [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+	<pluginVersion><![CDATA[3.2.0.201303060654-RELEASE]]></pluginVersion>
+	<configSuffixes>
+		<configSuffix><![CDATA[xml]]></configSuffix>
+	</configSuffixes>
+	<enableImports><![CDATA[false]]></enableImports>
+	<configs>
+	</configs>
+	<configSets>
+	</configSets>
+</beansProjectDescription>
```
Comment (amilcar-infante) [inline topic7/.springBeans:13]: This file shouldn't be in the commit, please add it to .gitignore

File: topic7/src/main/java/models/Teacher.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	public Teacher(long teacherId) {
+		this.teacherId = teacherId;
+	}
+
+	public Teacher(String name, String lastName2) {
+		this.firtName = name;
+		this.lastName = lastName2;
+	}
+
+	public Long getteacherId() {
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/Teacher.java:36]: By convention the name of the methods should be getTeacherId

File: topic7/src/main/java/controllers/TeacherController.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+public class TeacherController {
+
+	@Autowired
+	private TeacherDao _TeacherDao;
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/TeacherController.java:15]: By convention the name of the variables needs to be using camelCase (i.e: teacherDao)

File: topic7/pom.xml (hunk 1 above)
Comment (amilcar-infante) [inline topic7/pom.xml:9]: This one right? https://github.com/netgloo/netgloo-blog/blob/master/spring-boot-mysql-springdatajpa-hibernate/pom.xml

File: topic7/pom.xml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    </dependency>
+    <dependency>
+      <groupId>mysql</groupId>
+      <artifactId>mysql-connector-java</artifactId>
+    </dependency>
+  </dependencies>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <start-class>netgloo.Application</start-class>
```
Comment (amilcar-infante) [inline topic7/pom.xml:39]: The start class that you are using in this properties is not related with you project, it should be "test/java/topic7.Application"

File: topic7/pom.xml (hunk 1 above)
Comment (amilcar-infante) [inline topic7/pom.xml:9]: The artifactId is generally the name that the project is known by. In your case it should be topic7

File: topic7/pom.xml [hunk 10]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project
+  xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>netgloo</groupId>
```
Comment (amilcar-infante) [inline topic7/pom.xml:8]: The groupId is generally unique amongst an organization or a project. The group acts much like the Java packaging structure does in an operating system. In this exercise if you had declared a package "com,mhaure" that should be the value to be configure here.

File: topic7/src/main/resources/application.properties [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+# Specify the DBMS
+spring.jpa.database = MYSQL
+
+# Show or not log for each sql query
+spring.jpa.show-sql = true
+
+# Hibernate settings are prefixed with spring.jpa.hibernate.*
+spring.jpa.hibernate.ddl-auto = update
+spring.jpa.hibernate.dialect = org.hibernate.dialect.MySQL5Dialect
+spring.jpa.hibernate.naming_strategy = org.hibernate.cfg.ImprovedNamingStrategy
```
Comment (amilcar-infante) [inline topic7/src/main/resources/application.properties:16]: Looks good! Did you validate that the database is reachable at localhost:3306?

File: topic7/src/main/java/controllers/CourseController.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	@ResponseBody
+	public String findAll() {
+		try {
+			_CourseDao.findAll();
+		} catch (Exception ex) {
+			return "Error finding:" + ex.toString();
+		}
+		return "Courses succesfully founded!";
+	}
+}
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:51]: It´s a good practice, as we discussed in topic 6, to use the following layers: controllers - services - DAO. In this case isn't 100% necessary, but it is recommended to call the service from the controller and let the service to call the DAO.

File: topic7/src/main/java/controllers/CourseController.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+public class CourseController {
+
+	@Autowired
+	private CourseDao _CourseDao;
+
+	@RequestMapping("/create")
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:17]: Just to know more about it, please read this post http://www.journaldev.com/3358/spring-mvc-requestmapping-annotation-example-with-controller-methods-headers-params-requestparam-pathvariable

File: topic7/src/main/java/controllers/CourseController.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+			return "Error creating the user: " + ex.toString();
+		}
+		return "Course succesfully created!";
+	}
+
+	@RequestMapping("/delete")
+	@ResponseBody
+	public String delete(long id) {
+		try {
+			Course user = new Course(id);
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:33]: I don't understand why you are returning a course by Id and set it in a variable user

File: topic7/src/main/java/controllers/CourseController.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	@RequestMapping("/findAll")
+	@ResponseBody
+	public String findAll() {
+		try {
+			_CourseDao.findAll();
+		} catch (Exception ex) {
+			return "Error finding:" + ex.toString();
+		}
+		return "Courses succesfully founded!";
+	}
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:50]: You are creating, removing and finding all the courses. How could you be returning all the students related to a course?, and the teacher assigned to the course?

File: topic7/src/main/java/controllers/StudentController.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+	public String findAll() {
+		try {
+			_studentDao.findAll();
+		} catch (Exception ex) {
+			return "Error finding:" + ex.toString();
+		}
+		return "Students succesfully founded!";
+	}
+
+}
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/StudentController.java:52]: How could you get all the courses related to a student?

File: topic7/src/main/java/controllers/TeacherController.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+	public String findAll() {
+		try {
+			_TeacherDao.findAll();
+		} catch (Exception ex) {
+			return "Error finding:" + ex.toString();
+		}
+		return "Teachers succesfully founded!";
+	}
+
+}
```
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/TeacherController.java:52]: Could you add a method/request to return all the courses assigned to a teacher?

File: topic7/src/main/java/models/Course.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.OneToMany;
+import javax.persistence.Table;
+import javax.validation.constraints.NotNull;
+
+@Entity
+@Table(name = "course")
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/Course.java:12]: Please validate if the name of the table is "course" in lowercase and if the fields are being correctly mapped (it that's not the case you could use @Column(name = "<fied name>"), when <field name> could be, for example, course_name

File: topic7/src/main/java/models/Course.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+	public int getTeacherAssigned() {
+		return teacherAssigned;
+	}
+
+	public void setTeacherAssigned(int teacherAssigned) {
+		this.teacherAssigned = teacherAssigned;
+	}
+
+}
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/Course.java:86]: Looks good!

File: topic7/src/main/java/models/CourseDao.java [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+import javax.transaction.Transactional;
+
+import org.springframework.data.repository.CrudRepository;
+
+@Transactional
+public interface CourseDao extends CrudRepository<Course, Long> {
+
+	public Course findByName(String name);
+
+}
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/CourseDao.java:12]: Looks good!!! Why are you using CrudRepository?

File: topic7/src/main/java/models/Notes.java [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+	public Long getStudentId() {
+		return studentId;
+	}
+
+	public Long getCourseId() {
+		return courseId;
+	}
+
+}
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/Notes.java:90]: Looks good!

File: topic7/src/main/java/models/NotesDao.java [hunk 22]
```diff
@@ -0,0 +1,12 @@
+package models;
+
+import javax.transaction.Transactional;
+
+import org.springframework.data.repository.CrudRepository;
+
+@Transactional
+public interface NotesDao extends CrudRepository<Notes, Long> {
+
+	public Notes findByStudentNameAndLastName(String name, String lastName);
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/NotesDao.java:10]: How is defined the query to be used by this method?

File: topic7/src/main/java/models/Schedule.java [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+
+	public String getHourEnd() {
+		return hourEnd;
+	}
+
+	public void setHourEnd(String hourEnd) {
+		this.hourEnd = hourEnd;
+	}
+
+}
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/Schedule.java:67]: :+1:

File: topic7/src/main/java/models/Teacher.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.AUTO)
+	private Long teacherId;
+
+	@NotNull
+	private String dateOfBirth;
+
+	@NotNull
+	private String firtName;
```
Comment (amilcar-infante) [inline topic7/src/main/java/models/Teacher.java:22]: Review the name of the variables :)

File: topic7/src/test/java/topic7/Application.java [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+
+		StudentController studentController = new StudentController();
+
+		studentController.create("1", "1");
+		studentController.create("2", "2");
+		studentController.create("3", "3");
+		studentController.create("4", "4");
+		studentController.create("4", "4");
+
+		System.out.println(studentController.findAll());
```
Comment (amilcar-infante) [inline topic7/src/test/java/topic7/Application.java:34]: It would be nice to test the entire implementation, but I understand why you did only these tests

**Comment by amilcar-infante** [comment]
Good work Matias!!! Please review my comments

File: topic7/src/main/java/models/Teacher.java (hunk 24 above)
Comment (matihaure) [inline topic7/src/main/java/models/Teacher.java:22]: :+1:

File: topic7/src/main/java/models/NotesDao.java (hunk 22 above)
Comment (matihaure) [inline topic7/src/main/java/models/NotesDao.java:10]: its not defined ! becouse i didnt use it

File: topic7/src/main/java/models/CourseDao.java (hunk 20 above)
Comment (matihaure) [inline topic7/src/main/java/models/CourseDao.java:12]: In case i need some other options ( Create , Remove , Update , FindAll)

File: topic7/src/main/java/models/Course.java (hunk 18 above)
Comment (matihaure) [inline topic7/src/main/java/models/Course.java:12]: all rigth !

File: topic7/src/main/java/controllers/CourseController.java (hunk 13 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/CourseController.java:17]: :+1:

File: topic7/src/main/java/controllers/CourseController.java (hunk 12 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/CourseController.java:51]: i had this comment in my mind for topic 8 !

File: topic7/src/main/resources/application.properties (hunk 11 above)
Comment (matihaure) [inline topic7/src/main/resources/application.properties:16]: yes !

File: topic7/pom.xml (hunk 10 above)
Comment (matihaure) [inline topic7/pom.xml:8]: i know , i forgot to change it from the sample :(

File: topic7/src/main/java/models/Teacher.java (hunk 7 above)
Comment (matihaure) [inline topic7/src/main/java/models/Teacher.java:36]: :+1:

File: topic7/src/main/java/controllers/TeacherController.java (hunk 8 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/TeacherController.java:15]: :+1:

File: topic7/src/main/java/controllers/CourseController.java (hunk 5 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/CourseController.java:26]: okey !

File: topic7/src/main/java/controllers/CourseController.java (hunk 4 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/CourseController.java:45]: ups ! copy paste

File: topic7/pom.xml (hunk 1 above)
Comment (matihaure) [inline topic7/pom.xml:9]: :+1:

File: topic7/src/test/java/topic7/Application.java (hunk 25 above)
Comment (matihaure) [inline topic7/src/test/java/topic7/Application.java:34]: :+1:

File: topic7/src/main/java/controllers/StudentController.java (hunk 16 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/StudentController.java:52]: @Query("Select s from Notes n join n.student s where n.idCourse = ?")

I had not done that I did not reach the time, but would go this judgment on the method in DAO

File: topic7/src/main/java/controllers/CourseController.java (hunk 12 above)
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/CourseController.java:51]: :+1:

File: topic7/src/main/java/controllers/StudentController.java (hunk 16 above)
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/StudentController.java:52]: Please review the query in the comment

File: topic7/src/main/java/models/CourseDao.java (hunk 20 above)
Comment (amilcar-infante) [inline topic7/src/main/java/models/CourseDao.java:12]: Good!

File: topic7/src/main/java/models/NotesDao.java (hunk 22 above)
Comment (amilcar-infante) [inline topic7/src/main/java/models/NotesDao.java:10]: Please read point 3.3.2. Query creation in http://docs.spring.io/spring-data/jpa/docs/current/reference/html/ to have more information

File: topic7/src/main/java/controllers/CourseController.java (hunk 2 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/CourseController.java:20]: :+1:

File: topic7/.springBeans (hunk 6 above)
Comment (matihaure) [inline topic7/.springBeans:13]: :+1:

File: topic7/pom.xml (hunk 9 above)
Comment (matihaure) [inline topic7/pom.xml:39]: :+1:

File: topic7/src/main/java/controllers/CourseController.java (hunk 14 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/CourseController.java:33]: copy paste error , sry

File: topic7/src/test/java/topic7/Application.java (hunk 25 above)
Comment (matihaure) [inline topic7/src/test/java/topic7/Application.java:34]: :+1:

File: topic7/src/main/java/controllers/StudentController.java (hunk 16 above)
Comment (matihaure) [inline topic7/src/main/java/controllers/StudentController.java:52]: //Students by course

@Query("Select s from Notes n join n.student s where n.courseId = ?")
List<Student> findStudentsByCourse(Integer courseId);

//Courses by student

@Query("Select c from Notes n join n.course c where n.registrationNumber = ?")
List<Course> findCoursesByStudent(Integer registrationNumber);

File: topic7/src/main/java/models/NotesDao.java (hunk 22 above)
Comment (matihaure) [inline topic7/src/main/java/models/NotesDao.java:10]: :+1:

File: topic7/src/main/java/controllers/StudentController.java (hunk 16 above)
Comment (amilcar-infante) [inline topic7/src/main/java/controllers/StudentController.java:52]: Sorry Matias but I don't understand why are you using "n.students" in the join and where are you declaring the fields using in the Join (On clause) 
SELECT <select_list> 
FROM Table_A A
INNER JOIN Table_B B
ON A.Key = B.Key

Comment (matihaure) [inline topic7/src/main/java/controllers/StudentController.java:52]: @Query(SELECT p FROM Notes n INNER JOIN Student s ON n.idCourse = s.idCourseWHERE n.idCourse = ?)

Comment (amilcar-infante) [inline topic7/src/main/java/controllers/StudentController.java:52]: Exactly!

**Comment by amilcar-infante** [comment]
Good work Matias!!!
